feat: Add support for official Docker images of Fixer (#7555)
PHP-CS-Fixer/PHP-CS-Fixer: README.md

## Documentation

> ### Run with docker

Reviewer: pls put it after `### Installation` , I would still expect ppl to install first via Composer

Reviewer 2: Ah, good idea 👍. Also `Docker` with ucfirst, I missed it somehow 😅.

Author: Fixed it.

> You can take a ready built docker image to run ``php-cs-fixer``.

Reviewer: how will custom tokens act with docker image?
is it possible? if not out of the box, please describe how (or claim, in docs, lack of support)

Reviewer 2: If you mean custom rules, then these will work normally because you need to mount whole project as a volume inside the Docker container, to run Fixer against the code. So effectively you mount it with your code and vendors. It's basically out of the scope of the PR because it's about using Docker properly and we're not here to teach it 😉.

Reviewer: awesome. I was afraid some magic may be needed or not possible.
if business as usual, all great.

```console
docker run -v $(pwd):/var/www ghcr.io/php-cs-fixer/php-cs-fixer:latest fix src
```

or integrate as check into gitlab-ci like this

```yaml
php-cs-fixer:
  image: ghcr.io/php-cs-fixer/php-cs-fixer:3.41-php8.2
  script:

>     php-cs-fixer fix --diff --dry-run --format=txt src

Reviewer 2: Is it actually a correct command? Right now containers use `php-cs-fixer` as an entrypoint, so executed command should be a Fixer's command:

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/30897c63-9c27-4911-bee6-aa4ef171e071)

It would be great to utilise entrypoint like [`tini`](https://github.com/krallin/tini) to make it possible to run native commands on the container too. You can see how it's done in [Composer's builds](https://github.com/composer/docker/blob/main/2.6/docker-entrypoint.sh), the difference in usage:

![image](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/assets/600668/adb7e981-0a48-4a7d-a1b7-113ba4e1455f)

PS. As you can see on screenshots, my platform (MacOS) is not supported natively, is it possible to make multi-arch build?

Reviewer 2: @hgiesenow please keep in mind that I can't merge this without @keradus' decision, so if you're afraid about spending your time "for nothing", maybe wait for his comment 🙂. It can be in Q1.2024 though.

Author: > Is it actually a correct command? 
Yes it is, as gitlab ignores the entrypoint in favor of a shell, running the script. So this will work.

> It would be great to utilise entrypoint like [tini](https://github.com/krallin/tini) to make it possible to run native commands on the container too
I don't see the point in this yet. If you want to run a different command you can do it by overwriting the entrypoint (`docker run --entrypoint=""`). I would rather stick to docker standards then adding more "bloat" to it.

> PS. As you can see on screenshots, my platform (MacOS) is not supported natively, is it possible to make multi-arch build?
I found a documentation for that, but that would add much more complexity, so I would suggest this for a later iteration. I think it's "only" a performance topic.

Reviewer 2: > Yes it is, as gitlab ignores the entrypoint in favor of a shell, running the script. So this will work.

👍 

> I don't see the point in this yet. If you want to run a different command you can do it by overwriting the entrypoint (docker run --entrypoint=""). I would rather stick to docker standards then adding more "bloat" to it.

Yes, but this is the worse UX, people not familiar with Docker enough will scratch their head why `docker run [...] sh` does not work 🙂. Adding `tini` keeps default action as-is, but enables user-friendly access to other commands. Not a blocker in any way, though - it can be added later.

> I found a documentation for that, but that would add much more complexity, so I would suggest this for a later iteration. I think it's "only" a performance topic.

Yes, it's a performance thing mostly and is not a blocker too 🙂.

```

There are different tags for each stability and php version with syntax `<php-cs-fixer-version>-php<php-version>`. For example:

* `3.41.1-php8.1`
* `3.41-php8.2`
* `3` (uses the latest php version)
* `latest` (latest stable cs-fixer and latest stable php version)

> * `edge-php7.4` (current build from main with old php version)

Author: I need to delete these parts either way, when we are not building against the latest php version.

For the `1`you did look in the wrong job output I guess. It is built for the php8.3 version only https://github.com/elbformat/PHP-CS-Fixer/actions/runs/7249172660/job/19746615459

For more details and other installation methods, see
[installation instructions](./doc/installation.rst).

> ### Run with Docker

Reviewer: ```suggestion
### Run with Docker (early access)
```

only idea. I'm not sure if we confident all will work great with very first version of it, or maybe we mention to users it's early access here and they could experience some imperfections

Reviewer 2: I am pretty sure it will work as expected, because images are currently built [in the @hgiesenow's fork](https://github.com/elbformat/PHP-CS-Fixer/actions/workflows/docker.yml) and you can pull them from the registry and verify 🙂.

Author: I would also recommend not to add this. Otherwise we would need to define, when the early access ends and why. I think we should instead iteratively improve it, as soon as there are issues appearing.

> You can take a ready built docker image to run ``php-cs-fixer``.

Reviewer 2: ```suggestion
You can use pre-built Docker images to run ``php-cs-fixer``.
```

```console

> docker run -v $(pwd):/code ghcr.io/php-cs-fixer/php-cs-fixer:latest fix src

Reviewer 2: We can't use `latest` here.

```suggestion
docker run -v $(pwd):/code ghcr.io/php-cs-fixer/php-cs-fixer:$FIXER_VERSION fix src
```

> docker run -v $(pwd):/code ghcr.io/php-cs-fixer/php-cs-fixer:$FIXER_VERSION fix src

Reviewer: we could also say `php-cs-fixer:^3-php8.3 fix src` and avoid teaching user about extra env var

or have it like `${FIXER_VERSION:-^3-php8.3}`.

Reviewer 2: Can we improve it later to not block this PR longer?

Reviewer: it's one-liner to apply, even from diff view...
if we insist, we can externalise this to separated PR

Reviewer 2: It's only because I already wrote "one step before merge" (twice) and from @hgiesenow's perspective it may be frustrating to do changes back and forth, depending on who comments 😅. I can address your concerns later, these are not crtitical.

Author: I added a default as suggested. That will improve acceptance and keeps it flexibel.

```

Or integrate as check into Gitlab CI like this:

```yaml
php-cs-fixer:

>   image: ghcr.io/php-cs-fixer/php-cs-fixer:3.41-php8.2

Reviewer 2: ```suggestion
  image: ghcr.io/php-cs-fixer/php-cs-fixer:$FIXER_VERSION
```

  script:
    php-cs-fixer fix --diff --dry-run --format=txt src

> ```

Reviewer: I would like to keep readme minimalistic, docker usage info is great, but integrating with single specific CI, I would not put it in readme at all.

Drop it, move to https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/blob/master/doc/installation.rst or create specific doc file for CI integration (then I would even prioritise github over gitlab, but OK to have both, or none)

Reviewer 2: Again - can we improve it in consequent PR?

Reviewer: I see no value in externalising all the comments. If we have typo in PR, you would just fix it in PR, not externalise to new PR, right? :)

> There are different tags for each stability and php version with syntax `<php-cs-fixer-version>-php<php-version>`. For example:
> 
> * `3.41.1-php8.1`
> * `3.41-php8.2`
> * `3-php7.4`

Reviewer 2: ```suggestion

`$FIXER_VERSION` used in example above is an identifier of a release you want to use, which is based on Fixer and PHP versions combined. There are different tags for each Fixer's SemVer level and PHP version with syntax `<php-cs-fixer-version>-php<php-version>`. For example:

* `3.47.0-php7.4`
* `3.47-php8.0`
* `3-php8.3`
```

Let's change all `3.41` to `3.47` because docs should not point to version that won't be available in the registry. @keradus please let's wait with `3.47.0` until we merge this 🙂. It's not critical, though, we can always fix it later.
